scipy/cluster/hierarchy.py

```diff
@@ -836,6 +856,116 @@ def pre_order(self, func=(lambda x: x.id)):
 _cnode_type = type(ClusterNode)
 
 
+def _order_cluster_tree(Z):
+    """
+    Returns clustering nodes in bottom-up order by distance.
+
+    Parameters
+    ----------
+    Z : scipy.cluster.linkage array
+        The linkage matrix.
+
+    Returns
+    -------
+    nodes : list
+        A list of ClusterNode objects.
+    """
+    q = deque()
+    tree = to_tree(Z)
+    q.append(tree)
+    nodes = []
+
+    while q:
+        node = q.popleft()
+        if not node.is_leaf():
+            bisect.insort_left(nodes, node)
+            q.append(node.get_right())
+            q.append(node.get_left())
+    return nodes
+
+
+def cut_tree(Z, n_clusters=None, height=None):
+    """
+    Given a linkage Z, return the cut tree.
+
+    Parameters
+    ----------
+    Z : scipy.cluster.linkage array
+        The linkage matrix.
+    n_clusters : float, array-like, or None
+        Number of clusters in the tree at the cut point.
+    height : float, array-like, or None
+        The height at which to cut the tree. Only possible for ultrametric
```

Comment:
what's an ultrametric tree? I don't know that much about clustering, but I suspect this deserves an explanation